workers/unity/Packages/com.improbable.gdk.core/Utility/EntityTemplate.cs

```diff
@@ -1,42 +1,272 @@
+        public TSnapshot? GetComponent<TSnapshot>() where TSnapshot : struct, ISpatialComponentSnapshot
         {
-            if (hasBeenUsed)
+            if (entityData.TryGetValue(DynamicSnapshot.GetSnapshotComponentId<TSnapshot>(), out var snapshot))
             {
-                throw new InvalidOperationException("Cannot create an entity using the same EntityTemplate twice.");
+                return (TSnapshot) snapshot;
             }
 
-            hasBeenUsed = true;
-            return template;
+            return null;
+        }
+
+        /// <summary>
+        ///     Checks if a component snapshot is stored in the EntityTemplate.
+        /// </summary>
+        /// <param name="componentId">The component id to check.</param>
+        /// <returns>True, if the component snapshot exists, false otherwise.</returns>
+        public bool HasComponent(uint componentId)
+        {
+            return entityData.ContainsKey(componentId);
+        }
+
+        /// <summary>
+        ///     Checks if a component snapshot is stored in the EntityTemplate.
+        /// </summary>
+        /// <typeparam name="TSnapshot">The type of the component snapshot.</typeparam>
+        /// <returns>True, if the component snapshot exists, false otherwise.</returns>
+        public bool HasComponent<TSnapshot>() where TSnapshot : struct, ISpatialComponentSnapshot
+        {
+            return HasComponent(DynamicSnapshot.GetSnapshotComponentId<TSnapshot>());
         }
 
-        public void Dispose()
+        /// <summary>
+        ///     Attempts to overwrite a component snapshot in the EntityTemplate with another one of the same type.
+        /// </summary>
+        /// <param name="snapshot">The component snapshot that will be inserted into the EntityTemplate.</param>
+        /// <typeparam name="TSnapshot">The type of the component snapshot.</typeparam>
+        /// <exception cref="InvalidOperationException">
+        ///     Thrown if there is no component snapshot of type <see cref="TSnapshot"/> to replace.
+        /// </exception>
+        public void ReplaceComponent<TSnapshot>(TSnapshot snapshot) where TSnapshot : struct, ISpatialComponentSnapshot
         {
-            if (hasBeenUsed)
+            if (!entityData.ContainsKey(snapshot.ComponentId))
             {
-                return;
+                throw new InvalidOperationException("Can only replace a component if one already exists in an EntityTemplate.");
             }
 
-            hasBeenUsed = true;
+            entityData[snapshot.ComponentId] = snapshot;
+        }
+
+        /// <summary>
+        ///     Removes a component snapshot from the EntityTemplate, if it exists.
+        /// </summary>
+        /// <typeparam name="TSnapshot">The type of the component snapshot.</typeparam>
+        public void RemoveComponent<TSnapshot>() where TSnapshot : struct, ISpatialComponentSnapshot
```

Comment:
Would it be preferable to have this is a `SetComponent` function rather than separate add and replace?
Also in our current use cases would it be better to have this return a copy but with the replacement?

Reply:
This was meant to be on replace rather than remove

Reply:
Renamed `ReplaceComponent` -> `SetComponent` and removed the check as discussed offline.